### PR TITLE
#1893/#1894: fail-closed configstore constructor + pkg/fsatomic durable writers

### PR DESCRIPTION
**Pull request by psaab**
## Summary

Configstore durability lane from Codex audit `codex-review-008`:

- **#1893** — `configstore.New()` logged "falling back to file-only" on
  `NewDB` failure, stored a nil `*DB`, and the daemon nil-deref panicked at
  the first `Load()`/`writeActive()`. There is no file-only backend. The
  constructor now **fails closed** with a precise error
  (`config db <dir> unusable: ... refusing to boot without config
  persistence`); `daemon.New` returns the error and `cmd/xpfd` exits 1.
- **#1894** — no fsync anywhere on durable state. New package
  **`pkg/fsatomic`** provides `WriteFileAtomic` (temp+fchmod+close-check+
  rename, namespace atomicity only, NO fsync) and `WriteFileDurable`
  (adds temp-fd fsync before close + symlink-resolved parent-dir fsync
  after rename) plus `SyncDir`, with `WithPreserveExisting` /
  `WithResolveSymlinks` options lifted mechanically from FRR's
  reviewed-in-#1883 `atomicWriteFile`.

## Migration table (persistence class per site)

| Site | Class | Writer |
|---|---|---|
| `configstore/db.go writeTree` (active.json) | DurableState | `WriteFileDurable` 0644 — backs the #1799 persist-before-promote contract |
| `configstore/crypto.go readOrCreateMasterKey` | DurableState | `WriteFileDurable` 0600 — loss = undecryptable config |
| `configstore/store.go saveRollbackFiles` | slot 1 DurableState, slots 2..N Atomic | one trailing `SyncDir` covers shuffle + stale-slot unlinks |
| `configstore/store.go SaveRescueConfig` | DurableState | `WriteFileDurable` 0644 |
| `configstore/store.go ArchiveConfig` | AtomicGeneratedConfig | `WriteFileAtomic` 0644 |
| `dhcp/dhcp.go saveDUID` | DurableState | `WriteFileDurable` 0644 |
| `frr/manager.go atomicWriteFile` | DurableState | thin wrapper, `WithPreserveExisting`+`WithResolveSymlinks` — gains the missing dir fsync |
| `ipsec/ipsec.go Apply` | AtomicGeneratedConfig | `WriteFileAtomic` 0600 |
| `dhcpserver writeKeaConfig` | AtomicGeneratedConfig | `WriteFileAtomic` 0644 |
| `networkd writeIfChanged` | AtomicGeneratedConfig | `WriteFileAtomic` 0644 — zero fsync on apply path |
| `networkd restoreSlowPathRPFilter` | BestEffortKernelKnob | direct `os.WriteFile` stays (procfs) |
| `configstore/journal.go` | untouched | #1896 owns the journal redesign |

Commit path pays 2 file fsyncs + 3 dir fsyncs total (operator-paced,
bounded). Hot apply paths (networkd/swanctl/Kea) pay zero fsync by class.
An AST-based canary test (`pkg/fsatomic/canary_test.go`) enforces the
allowlist of remaining direct `os.WriteFile` calls in migrated packages.

## Plan + round-1 adversarial review

Plan v2 (`docs/pr/1893-configstore-durability/plan.md`) adopts all
sustained round-1 findings: Codex task-mqal7v2u-8adha1 (PLAN-NEEDS-MAJOR,
7 findings sustained) + AGY adversarial-review-mqal37pw-ehzagk
(PLAN-NEEDS-MAJOR; A4 "hallucinated #1799 machinery" REFUTED — it audited
a stale checkout). Key adoptions: slot-1-durable rollback design (one
trailing SyncDir instead of ~50 fsyncs), ENOTDIR constructor test
(root-proof), AST canary, crash-leaked temp sweep in `NewDB`.

## Validation

- `go build ./...` = 0; full `go test ./...` = 0 (unmasked); `go test
  -race -count=1 ./pkg/configstore/ ./pkg/fsatomic/` = 0 — re-run after
  rebase onto post-#1897 master.
- Per-stage injected-failure tests in fsatomic (every seam: create,
  write, chmod, chown, sync, close, rename, dir-open, dir-sync); #1893
  constructor ENOTDIR test asserts error-not-panic.
- Live fail-closed validation on the loss userspace cluster: results to
  be posted as a PR comment.

Closes #1893
Closes #1894

### CODE REVIEW

**Comment by psaab** [comment]
## Live fail-closed validation (loss userspace cluster, lock cell "1893 fail-closed validation")

Branch build `d02276728` rolling-deployed to both nodes (`Software version: ...-gd02276728` on node0 and peer).

**Normal operation (fw0):** boot clean (no panics), commit probe `commit complete: 1 statement(s) changed`, revert commit clean, `show system rollback` lists entries, `.configdb/active.json` present, **no leaked `.*.tmp-*` temps** after boot+commits.

**Fail-closed (fw1):** stopped xpfd, replaced `/etc/xpf/.configdb` with a regular file (MkdirAll -> ENOTDIR), started xpfd. Journal — the precise constructor error, exit 1, NO boot-then-panic, NO "falling back to file-only":

```
Jun 12 07:58:28 xpf-userspace-fw1 xpfd[466551]: xpfd: config store: config db /etc/xpf/.configdb unusable: create db dir: mkdir /etc/xpf/.configdb: not a directory (no file-only fallback exists; refusing to run without config persistence)
Jun 12 07:58:28 xpf-userspace-fw1 systemd[1]: xpfd.service: Main process exited, code=exited, status=1/FAILURE
...
Jun 12 07:58:34 xpf-userspace-fw1 systemd[1]: xpfd.service: Failed with result 'start-limit-hit'.
```

`ActiveState=failed NRestarts=5 ExecMainStatus=1` — identical precise error on every restart attempt until systemd's start limit; no nil-deref panic anywhere in the window.

**Recovery:** restored `.configdb`, `systemctl reset-failed` + restart -> fw1 `active`, `xpf-userspace-dp` helper running. CoS re-applied (`apply-cos-config: atomic commit + verification OK`, all 12 queues present). Final: both nodes `active`, node0 primary, `Takeover ready: yes`, RG0 failover count 0. Cluster left on the branch build, healthy.

Gates after rebase onto post-#1897 master: `go build ./...` = 0, `go test ./...` = 0 (unmasked), `go test -race -count=1 ./pkg/configstore/ ./pkg/fsatomic/` = 0.

**Comment by psaab** [comment]
## Review status — MERGE-READY (3-of-4 fallback)

**Code round 1 (d02276728):** Codex task-mqamuw8f-l8iivi — NEEDS-CHANGES, one High: durable-state files written into directories the writer itself just created were not durable (`WriteFileDurable` fsyncs the file's parent, not the parent's entry in *its* parent), so on first boot `.configdb` — master.key and active.json included — could vanish whole after a power cut; same for the DHCP DUID state dir. **Fixed in cfc353613**: new `fsatomic.MkdirAllDurable` (fsync every newly-created level + the deepest pre-existing ancestor; zero fsyncs when the path exists), migrated `configstore.NewDB` + `dhcp.saveDUID`, `TestMkdirAllDurable` pins the fsync counts.

**Code round 2 (bf613888c):** Codex task-mqanzp5j-rvqi1m — **MERGE-READY, findings none** (first-boot power-loss trace closed end-to-end; missing-level walk sound under races/ENOTDIR/relative paths/root; no other durable-class runtime-created parent on plain MkdirAll).

**Claude SMR (in-conversation):** MERGE-READY. Power-loss stage trace verified (write -> fchmod -> fchown -> fsync(tmp) -> close-check -> rename -> fsync(resolved parent)); master.key loss scenario closed (durable write + durable dir creation); fail-closed boot error message live-validated. Noted Low, non-blocking: crash-leaked `.xpf.conf.N.tmp-*` temps in /etc/xpf are not swept at boot (only `.configdb` is) — rare commit-paced crash window, documented class in the fsatomic README.

**AGY:** two runs (adversarial-review-mqamn437-f3k3lx, retry adversarial-review-mqan33cv-17vcl4) both degenerate — status "succeeded" with 0-byte result and log truncated at invocation (AGY-side outage). One-concise-retry protocol exhausted; its plan-round-1 review (adversarial-review-mqal37pw-ehzagk) did engage substantively and its sustained findings are in the shipped design.

**Copilot:** quota-blocked — initial attempt returned the quota error as its review; documented retries at 08:01Z, 08:28Z, 08:45Z produced no new review. 3 documented retries -> 3-of-4 fallback per protocol.

**Final-head redeploy:** cluster rolling-redeployed to 99dfd9206 (`Software version: ...-g99dfd9206` both nodes), no panics, CoS re-applied (`atomic commit + verification OK`), node0 primary, `Takeover ready: yes`. Gates at head: build 0 / full tests 0 / `-race -count=1` configstore+fsatomic 0.

Trail: plan + round records in `docs/pr/1893-configstore-durability/` (plan.md, reviewer-ids.md). Awaiting parent smoke + merge.

**Comment by psaab** [comment]
Gate completed post-AGY-recovery: AGY MERGE-READY (adversarial-review-mqb743ow — all five hostile checks validated; one minor symlink-chain observation recorded, not a config-path scenario) + Codex r2 + SMR; Copilot quota ×3 documented. Final smoke (merged head): both nodes boot clean over stored config through the new durable writers; commit history intact; P12R line rate; journals clean. Live fail-closed validation (ENOTDIR refusal) captured earlier on this PR.
